### ISSUE TITLE
Introduce RegEx before "Check for Palindrome" problem

### ISSUE THREAD

**Issue by alayek** [FreeCodeCamp/FreeCodeCamp#9232]
[Check for Palindrome](https://www.freecodecamp.com/challenges/check-for-palindromes) continues to be difficult to surmount for beginners in JavaScript.

This challenge has two parts to it:
- Process the input string `str` to remove non-alphanumeric characters from it.
- Use some logic to find palindromes.

This is the third challenge in **Basic Algorithmic Challenges**, and the two that comes before this, introduce the camper to recursion and String-to-Array with `reverse()`, `split()`, and `join()`.

---

In this challenge, the camper needs to come up with a non-trivial regular expression, to remove non-alphanumeric characters from the input string `str`.

And here's the kicker - `\w` does correspond to alphanumeric characters [`0-9]` and `[a-z]` (and also `[A-Z]`, but not relevant to this challenge, because we do case insensitive matching); but it also includes underscore `_`. The test cases indicate you have to exclude underscore.

---

I recommend introducing the camper to regular expression, with 1-2 problems before that, and teach them how to form and test their RegEx. At the very least, they should be able to use something like [regex101](https://regex101.com/).

Ideally, I would want a full-fledged curriculum dedicated only to RegEx. I have seen even experienced campers fear regular expressions (I am not experienced, but I fear it too!). If this issue is accepted, I can work with others on the core-team on getting a curriculum on RegEx on the live site.

I know that there are four problems that refers to regular expression:
- [Sift through text with regular expressions](https://www.freecodecamp.com/challenges/sift-through-text-with-regular-expressions)
- [Find numbers with regular expressions](https://www.freecodecamp.com/challenges/find-numbers-with-regular-expressions)
- [Find whitespaces with regular expressions](https://www.freecodecamp.com/challenges/find-whitespace-with-regular-expressions)
- [Invert regular expression matches](https://www.freecodecamp.com/challenges/invert-regular-expression-matches-with-javascript)

But none of them makes the camper comfortable forming non-trivial complex regular expressions. And in later challenges, like US telephone numbers, regular expression makes the difference between solved and unsolved. 

Besides, these four problems are from optional part of the curriculum. In the helpful links section for the palindrome problem, we can at least point them to these.

@QuincyLarson @BerkeleyTrue  @SaintPeter @CodeNonprofit FYI.

**Comment by arun1595** [comment]
I agree with @alayek  to teach campers Regex. Its one of the powerful constructs that should be taught to campers. A whole set of Regex challenges in the FCC curriculum would be really exciting.

**Comment by QuincyLarson** [comment]
@alayek Thank you for making this observation and putting so much thought into it. 

I agree with @arun1595 that we should expand our coverage of Regex in our curriculum. We just need to decide which concepts to cover beyond the ones we already do, and probably have more regex practice as well. So we're talking maybe an additional 15-20 challenges. 

Here is  a the website we used to use to teach regex. Perhaps we could aim to cover everything they cover? http://regexone.com/

**Comment by QuincyLarson** [comment]
@alayek I also think we should identify en easier regex challenge and move it before "Check for Palindrome" - any ideas?

**Comment by CodeNonprofit** [comment]
@alayek A comment from a frustrated camper, seen [here](https://medium.com/@bonnerviolin/i-just-finished-palindromes-last-night-after-ca-c719906dfd2f#.quc5k2v5c) suggested that he did end up supplementing the curriculum with http://regex101.com. We might look there for inspiration for the additional 15-20 challenges.

**Comment by alayek** [comment]
@CodeNonprofit the camper's comments were spot on. And I have already linked to [regex101](http://regex101.com/) in the original issue filing.

I use regex101 regularly, because it clearly points out what part of your regex is responsible what is being matched and skipped. It also has a list of suggestions on common regex tokens when you are creating the RegEx.

This is why I recommended that we have a solid 10-15 problem RegEx curriculum. The goal would be to make campers comfortable with regular expressions. There are some problems that can be insanely easy and maybe even one-liner, if RegEx is used.

**Comment by alayek** [comment]
@QuincyLarson 

---
TL;DR: None that I could think of at the moment. We need a problem that demonstrates and hand-holds till a camper finds a way to form their own complex reg-ex.

----
I checked the existing Regular Expression problems in optional sections, and these are my takeaways:
- First one tells you about what RegEx is
- Second one tells you about `\d`, special token for digits. And `+`, to denote one or more occurrences.
- Third one tells you about `\s`, special token for space.
- Fourth one tells you about inverting a token, for example `\s` to `\S` to pick everything other than what is matched by `\s`.

There is a video on regular expressions in the hikes section, but there are only two questions to it without much takeaway, IMHO.

Assuming someone didn't skip these optional parts, they would probably be able to find out `\w` for alpha-numeric characters. But, they would still miss out of **grouping** with `[]` or `(|)`. They would know `+`, but miss out on `*`, `?`. They would know about `//g` and `//i`, but miss out on `^`.

I would say the most important takeaway **not** present in those, is that you can form your non-trivial RegEx to match weird patterns that you need for your applications. I have faced situations at work where I need to validate database connection strings, work email etc.

Our wiki has one liner solutions for palindromes, caesar ciphers and many more problems - with Reg-Ex. Being able to form your own reg-ex, simplifies a lot of problems; much the same way `map()`, `reduce()`, and `filter()` do. It makes core pattern matching logic of your application concentrate within the RegEx, and make it more readable.

**Comment by erictleung** [comment]
@alayek great analysis of the palindromes challenge. I liked @QuincyLarson's idea of having like 15--20 challenge on regex. It is quite powerful and with tools like regex101.com can make the task less daunting.

Instead of lumping all the regex challenges into Basic JavaScript, should we just make an entire new section just on Regular Expressions? It seems like a self contained topic, similar in my mind like the Object Oriented and Functional Programming section.

**Comment by alayek** [comment]
@erictleung I was hoping we would do that! An entire section on Regular Expressions. From log parsers to antiviruses - a lot of things in the world run on regular expressions. Most importantly, regex is a _transferable skill_. 

JS Reg-Ex is not specific to just JS - with some tweaking, it would also work in Python or Perl or Ruby, or even Bash etc.

I am all for a separate section on regular expression. But we might want to keep it optional.

**Comment by SaintPeter** [comment]
@QuincyLarson Here is some proposed curriculum: https://github.com/FreeCodeCamp/FreeCodeCamp/issues/5794

**Comment by deesto** [comment]
I agree this is needed, but the tutorial at http://regexone.com does a pretty good job of teaching this, including an interactive tutorial and built-in example problems. Instead of reinventing that wheel, is there any way to refer to and/or partner with the site?

**Comment by alayek** [comment]
@deesto the tutorial on [regexone](http://regexone.com/) is cool! But it also has its deficiencies.

As I have mentioned above, it lets you continue without having to fully match Strings given. But the biggest problem I see with tutorials like regexone and similar others, is this: _it's isolated_

Let me explain what I meant by that: most people learn RegEx step by step (groupings, tokens, extracting) then they move on with their life. They would never use it unless there is absolutely no other way.

You won't use RegEx as often as you would use conditionals or loops or recursion in your code. Even when you would have to use, you would most likely Google for some existing ones and copy-paste for lack of time. 

Our aim is to form the RegEx curriculum in a way that it feels organic, with lot of practical use-cases. We would like to make the camper feel at-home using regular expressions.

As for reinventing the wheel, it's not a massive undertaking, compared to some of the things we have built so far and more so planning to build.

**Comment by deesto** [comment]
@alayek Understood; thanks for that. And I'd argue that RegEx is even more prevalent than being used as a last resort for solving a matching problem, especially in contexts outside of JS, so I am all for having it covered adequately. I just wouldn't want valuable effort wasted, and given your points and the proposed curriculum, that certainly wouldn't seem to be the case.

**Comment by DealPete** [comment]
@SaintPeter, in #5794, you provide a link to a branch, but it is down. Does the branch still exist?

**Comment by QuincyLarson** [comment]
Our expanded curriculum will have a regex section before the algorithm challenges.
